### PR TITLE
stats: add perf-annotation regexes used in stats tag extraction

### PR DESCRIPTION
**Pull request by jmarantz**
*Description*::
This is step 3 in the [plan to improve startup performance](https://goo.gl/YxzBaB). This is another step toward addressing #2373
- Added new server mode ```init_only``` which is used to fully initialize the server but not start the listener
- Tweaks slightly the perf output format as it looks better to left-justify the category and description.
- Instrumented the regex parsing in stats tag extraction, to learn what slows us down.  From the 10k example in https://github.com/envoyproxy/envoy/issues/2063:
```
Duration(us)  # Calls  Mean(ns)  StdDev(ns)  Min(ns)  Max(ns)  Category  Description                   
     1217329   680147      1789     1182.73      495   880932  re-miss   envoy.grpc_bridge_method      
     1205580   680147      1772     1227.24      486   924781  re-miss   cipher_suite                  
     1193745   680147      1755      1191.7      477   896321  re-miss   envoy.grpc_bridge_service     
      714594   680147      1050     1116.88      528   862691  re-miss   envoy.response_code           
      681802   680127      1002     1169.21      508   870749  re-miss   envoy.response_code_class     
      657579   680147       966     414.414      570   125234  re-miss   envoy.dynamo_partition_id     
      648601   680031       953     335.497      583    49374  re-miss   envoy.http_conn_manager_prefix
      647820   680147       952     345.378      565    39783  re-miss   envoy.dynamo_operation        
      646571   680147       950     347.486      486    22955  re-miss   envoy.ssl_cipher              
      645047   680147       948     336.928      504    30860  re-miss   envoy.mongo_callsite          
      640112   680147       941     600.817      535   399301  re-miss   envoy.dyanmo_table            
      639674   680147       940     344.334      521    56102  re-miss   envoy.http_user_agent         
      634744   680147       933     340.052      494    32314  re-miss   envoy.mongo_collection        
      630200   680147       926     1193.34      536   946328  re-miss   envoy.fault_downstream_cluster
      629252   680147       925      340.45      485    33107  re-miss   envoy.mongo_cmd               
      628669   680147       924     321.804      473    21084  re-miss   envoy.virtual_cluster         
      457053   680000       672     437.367      567   109544  re-match  envoy.cluster_name            
          82      116       710     284.499      462     1876  re-match  envoy.http_conn_manager_prefix
          21       20      1091     332.877      613     1646  re-match  envoy.response_code_class     
          17        9      1895     437.737     1524     2952  re-miss   envoy.listener_address        
           4        5       840     359.531      635     1481  re-match  envoy.listener_address        
```

*Risk Level*: Low

*Testing*:
//test/...
bazel build --compilation_mode=opt //source/exe:envoy-static --define=perf_annotation=enabled
bazel-bin/source/exe/envoy-static --mode init_only -c 10k_clusters.json

*Release Notes*: not yet -- will issue them when the remainder of the changes in this series roll out

### CODE REVIEW

**Comment by dnoe** [comment]
Let me know when this is ready to review and I'll assign some folks.

**Comment by jmarantz** [comment]
This entire PR  blocked on https://github.com/envoyproxy/envoy/pull/2623 as it depends on the existence of a working flow that initializes the server but exits without running anything.

However I could pull out just the new perf_annotation class and review that....

File: source/common/common/perf_annotation.cc [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+void PerfAnnotationContext::clear() {
+  PerfAnnotationContext* context = getOrCreate();
+#if PERF_THREAD_SAFE
+  std::unique_lock<std::mutex> lock(context->mutex_);
+#endif
+  context->duration_count_map_.clear();
+}
+
+PerfAnnotationContext* PerfAnnotationContext::getOrCreate() {
+  static PerfAnnotationContext* context = new PerfAnnotationContext();
```
Comment (mrice32) [inline source/common/common/perf_annotation.cc]: Is there any reason we can't just subclass the [ThreadSafeSingleton](https://github.com/envoyproxy/envoy/blob/master/source/common/singleton/threadsafe_singleton.h#L24) to keep this singleton implementation consistent with others like it across Envoy?

File: source/common/common/perf_annotation.h [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+/**
+ * Represents an operation for reporting timing to the perf system. Usage:
+ *
+ * f() {
+ *   PerfOperation perf_op;
+ *   computeIntensiveWork();
+ *   perf_op.record("category", "description");
+ * }
+ */
+class PerfOperation {
```
Comment (mrice32) [inline source/common/common/perf_annotation.h]: There don't appear to be any tests for this class below.  Can we add a few?  I think just using a second timer to ensure that the recorded time will be greater than some value will suffice.

**Comment by mrice32** [comment]
Forgot to give an overall comment :).  This is really cool work!  It might be interesting to explore adding histograms to this once we start using a persistent histogram implementation for stats.

**Comment by jmarantz** [comment]
Closing this temporarily as the perf annotation library is in a #2626; once that's in I'll re-open this with the application to stats regexes.  @mrice32 's comments will be applied to #2626.

**Comment by jmarantz** [comment]
OK this has now merged in #2626 and has only the application of it to the stats init problem.  PTAL.

File: source/common/common/perf_annotation.cc [hunk 3]
```diff
@@ -122,7 +122,9 @@ std::string PerfAnnotationContext::toString() {
   // Create format-strings to right justify each column, e.g. {:>14} for a column of width 14.
   std::vector<std::string> formats;
   for (size_t i = 0; i < num_columns; ++i) {
-    formats.push_back(absl::StrCat("{:>", widths[i], "}"));
+    // left-justify category & description, but right-justify the numeric columns.
+    const char* justify = (i < num_columns - 2) ? ">" : "<";
```
Comment (htuch) [inline source/common/common/perf_annotation.cc]: Nit: prefer `const std::string` over raw `const char*` where possible.

File: include/envoy/server/options.h [hunk 4]
```diff
@@ -28,7 +28,12 @@ enum class Mode {
    */
   Validate,
 
-  // TODO(rlazarus): Add a third option for "light validation": Mock out access to the filesystem.
+  /**
+   * Completely load and initialize the config, and then exit without running the listener loop.
+   */
+  InitOnly,
```
Comment (htuch) [inline include/envoy/server/options.h]: Is there a test for this?

Comment (jmarantz) [reply include/envoy/server/options.h]: Ah no, but if #2679 can get merged it'll be way easier to add one.

File: source/common/common/perf_annotation.cc (hunk 3 above)
Comment (jmarantz) [reply source/common/common/perf_annotation.cc]: Seems strictly slower but OK.  WDYT of absl::string_view for this instead?

File: include/envoy/server/options.h (hunk 4 above)
Comment (jmarantz) [reply include/envoy/server/options.h]: Added one....i resurrected the random component to the main_common_test base ID calculation as it seems you can't immediately open one of the required resources immediate after closing it in the previous test method.

**Pull request merged**